### DISCUSSION TITLE
Add preview of hex colors inside code chunk

### DISCUSSION THREAD

**Discussion by brunomioto** [quarto-dev/quarto-cli#5908] [Feature Requests]
### Description

As seen in the new version of RStudio (or Github itself), it would be great if we could have a preview of the colors when typing the hex code, like here: `#ffc127`.
When sharing plots' code, it's helpful to know what color each one is.

**Comment by mcanouil** [comment]
Could you clarify, where you want this exactly?

**Reply by brunomioto** [reply]
> Could you clarify, where you want this exactly?

In the code chunk itself
![image](https://github.com/quarto-dev/quarto-cli/assets/10459392/d046c0e2-d293-4e6d-a37a-3b365d0ac5ee)

**Reply by mcanouil** [reply]
So in the rendered documents?

**Reply by brunomioto** [reply]
Yes

**Comment by mcanouil** [comment]
Like this?

<table>
<tr><th>Quarto document</th><th>HTML</th></tr>
<tr>
<td>

````qmd
---
format: html
include-after-body:
  - text: |
      <script type="text/javascript">
      // Get all elements with a hex color code
      const elements = document.querySelectorAll('*:not(script):not(style)');
      for (let i = 0; i < elements.length; i++) {
        const element = elements[i];
        const text = element.innerHTML;
        const regex = /#([0-9a-fA-F]{3}|[0-9a-fA-F]{6})\b/g;
        const newText = text.replace(
          regex,
          `$&<span style="display: inline-block; color: $&;">&#9632;</span>`
        );
        element.innerHTML = newText;
      }
      </script>
---

The `#b22222` colour is a dark red colour, also known as firebrick.
It is a web safe colour.
The #b22222 colour hex could be obtained by blending #ff4444 with #640000.
Closest websafe colour is: #cc3333.

````

</td>
<td>
<img width="506" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/cc4e8612-01f1-4cba-a41e-0f43379ec4d1">
</td>
</tr>
</table>

**Reply by brunomioto** [reply]
Yes!

**Reply by mcanouil** [reply]
Note that this messes up code blocks.
Also adding colour, basically disable the ability to copy clean code.

**Reply by mcanouil** [reply]
I don't think this should be applied to code blocks at all.

**Reply by brunomioto** [reply]
Oh, you are right. So maybe we can add it to plain text only?

**Comment by mcanouil** [comment]
Here an updated version.

````qmd
---
format: html
include-after-body:
  - text: |
      <script type="text/javascript">
      // Author: Mickaël Canouil
      // Version: <1.0.0>
      // Description: Scan HTML to find hex colour code and add a coloured square next to it.
      // License: MIT
      function scanElements(element) {
        if (element.nodeType === Node.TEXT_NODE) {
          const regex = /#([0-9a-fA-F]{3}|[0-9a-fA-F]{6})\b/g;
          const newText = element.textContent.replace(
            regex,
            `$&<span style="display: inline-block; color: $&;">&#9632;</span>`
          );
          if (newText !== element.textContent) {
            const newElement = document.createElement('span');
            newElement.innerHTML = newText;
            element.parentNode.replaceChild(newElement, element);
          }
        } else if (element.nodeType === Node.ELEMENT_NODE && !element.classList.contains('sourceCode')) {
          for (let i = 0; i < element.childNodes.length; i++) {
            scanElements(element.childNodes[i]);
          }
        }
      }

      scanElements(document.body);
      </script>
---

The `#b22222` colour is a dark red colour, also known as firebrick.
It is a web safe colour.
The #b22222 colour hex could be obtained by blending #ff4444 with #640000.
Closest websafe colour is: #cc3333.

```{r}
my_colour <- "#b22222"
```
````

<img width="424" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/9101b56c-7e44-4899-895c-d7cb81ba8f50">

**Reply by mcanouil** [reply]
Using a Lua filter, it could be possible to do the same, but for LaTeX and HTML formats.

**Reply by brunomioto** [reply]
That's a great idea for a Quarto extension

**Reply by mcanouil** [reply]
Teasing, Lua filter extension incoming (still some tweaks, and LaTeX code to add).
<img width="1624" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/c72722fb-fb3a-45be-8b34-6c5469eaeeab">

**Reply by brunomioto** [reply]
Awesome! 
Just thinking here, is it possible to use the background color of hex code without breaking it like the little square?
<img width="643" alt="177159470-0a1610e6-a145-412e-9460-e4abdcf67cc1" src="https://github.com/quarto-dev/quarto-cli/assets/10459392/18f16523-f878-4fbe-bb73-4fee9d22cb4f">

**Reply by mcanouil** [reply]
Well, yes, but then you have the change to font colour to get sufficient contrast.
This require computation ...

**Comment by EmilHvitfeldt** [comment]
I have been working on a similar project last night!

right now it is written in `js` but I'm confident it could be translated to Lua. Still ironing out the kinks.

con:
- Doesn't automatically detect colors like @mcanouil's solution

pro:
- works for more types of colors
- doesn't have false positives

This is how it interfaces right now, but I would like to move to using shortcodes like `{{ colorbox rgb(10,100,0) }}`

```
Names one [Orange]{.color}

hex codes [#441100]{.color}

short hex codes [#123]{.color}

rgb [rgb(10, 100, 200)]{.color}

rgb with % [rgb(100%, 20%, 100%)]{.color}

hwb [hwb(135 0% 40%)]{.color}

hsl [hsl(240, 100%, 50%)]{.color}
```

![Screenshot 2023-06-13 at 3 31 49 PM](https://github.com/quarto-dev/quarto-cli/assets/14034784/d8833fca-5483-4551-8938-378c363ccada)

**Reply by mcanouil** [reply]
Shortcode is easy to add with my current filter.
For HTML, other formats can be handled, but it is going to be tricky for LaTeX.

I'am also going to try a code syntax for colours: `` `#b22222` `` which will produce `#b22222` or `` `rgb(178,34,34)` `` for `rgb(178,34,34)`.

**Comment by brunomioto** [comment]
Feature request: Copy to clipboard on click (maybe this could be an option inside the extension)
![image](https://github.com/quarto-dev/quarto-cli/assets/10459392/118f2ea6-1bce-46b6-af88-7242114b6861)

**Comment by mcanouil** [comment]
First release (v0.1.0): <https://github.com/mcanouil/quarto-preview-colour>.
Edit: added support for LaTeX in addition to HTML based formats in v0.2.0.

**Comment by mcanouil** [comment]
# :art: NEW: Quarto Preview Colour Extension v1.0.0! (*finally*)

:sparkles: Inline colour previews (code and/or text)
:sparkles: Hex, RGB, HSL, and HWB support
:sparkles: Click to copy colour codes (HTML)
:sparkles: HTML, LaTeX, Typst, DOCX & PPTX support

https://github.com/mcanouil/quarto-preview-colour
